// zero-blocks/core/navigation/nav-service.js
"use strict";
/**
 * Common service to be injected between components where those implementing common
 * ToggleView interface can register and toggle directives can call their methods.
 * TODO: Track currently active? Events?
 */
var NavigationService = (function () {
    function NavigationService() {
        this.navs = {};
    }
    NavigationService.prototype.add = function (id, navItem) {
        this.navs[id] = navItem;
    };
    NavigationService.prototype.remove = function (id) {
        delete this.navs[id];
    };
    NavigationService.prototype.get = function (id) {
        if (id) {
            return this.navs[id];
        }
    };
    NavigationService.prototype.toggle = function (id, fireEvents) {
        if (this.navs[id]) {
            return this.navs[id].toggle(fireEvents);
        }
        else {
            return Promise.reject("No ToggleView component found for id:" + id);
        }
    };
    NavigationService.prototype.open = function (id, fireEvents) {
        if (this.navs[id]) {
            return this.navs[id].open(fireEvents);
        }
        else {
            return Promise.reject("No ToggleView component found for id:" + id);
        }
    };
    NavigationService.prototype.close = function (id, fireEvents) {
        if (this.navs[id]) {
            return this.navs[id].close(fireEvents);
        }
        else {
            return Promise.reject("No ToggleView component found for id:" + id);
        }
    };
    return NavigationService;
}());
exports.NavigationService = NavigationService;
