// zero-blocks/button/button.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var IgButton = (function () {
    function IgButton(_el, _renderer) {
        this._el = _el;
        this._renderer = _renderer;
        this._type = 'flat';
        this._cssClass = 'ig-button';
    }
    Object.defineProperty(IgButton.prototype, "type", {
        set: function (value) {
            this._type = value || this._type;
            this._renderer.setElementClass(this._el.nativeElement, this._cssClass + "--" + this._type, true);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgButton.prototype, "color", {
        set: function (value) {
            this._color = value || this._el.nativeElement.style.color;
            this._renderer.setElementStyle(this._el.nativeElement, 'color', this._color);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgButton.prototype, "background", {
        set: function (value) {
            this._backgroundColor = value || this._backgroundColor;
            this._renderer.setElementStyle(this._el.nativeElement, 'background', this._backgroundColor);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgButton.prototype, "disabled", {
        set: function (val) {
            this._renderer.setElementClass(this._el.nativeElement, this._cssClass + "--disabled", !!val);
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Input('igButton'), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], IgButton.prototype, "type", null);
    __decorate([
        core_1.Input('igButtonColor'), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], IgButton.prototype, "color", null);
    __decorate([
        core_1.Input('igButtonBackground'), 
        __metadata('design:type', String), 
        __metadata('design:paramtypes', [String])
    ], IgButton.prototype, "background", null);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object), 
        __metadata('design:paramtypes', [Object])
    ], IgButton.prototype, "disabled", null);
    IgButton = __decorate([
        core_1.Directive({
            selector: '[igButton]',
            host: {
                'role': 'button'
            }
        }), 
        __metadata('design:paramtypes', [core_1.ElementRef, core_1.Renderer])
    ], IgButton);
    return IgButton;
}());
exports.IgButton = IgButton;
var ButtonModule = (function () {
    function ButtonModule() {
    }
    ButtonModule = __decorate([
        core_1.NgModule({
            declarations: [IgButton],
            exports: [IgButton]
        }), 
        __metadata('design:paramtypes', [])
    ], ButtonModule);
    return ButtonModule;
}());
exports.ButtonModule = ButtonModule;
