// zero-blocks/modal/modal.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var Modal = (function () {
    function Modal() {
        this.closeOnClick = true;
        this.onOpen = new core_1.EventEmitter();
        this.onClose = new core_1.EventEmitter();
        this.isOpen = false;
    }
    Modal.prototype.open = function () {
        if (this.isOpen) {
            return;
        }
        this.isOpen = true;
        this.onOpen.emit(this);
        this.modalEl.nativeElement.classList.add('ig-modal');
    };
    Modal.prototype.close = function () {
        if (!this.isOpen) {
            return;
        }
        this.isOpen = false;
        this.onClose.emit(this);
        this.modalEl.nativeElement.classList.remove('ig-modal');
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], Modal.prototype, "closeOnClick", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', Object)
    ], Modal.prototype, "onOpen", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', Object)
    ], Modal.prototype, "onClose", void 0);
    __decorate([
        core_1.ViewChild('modal'), 
        __metadata('design:type', core_1.ElementRef)
    ], Modal.prototype, "modalEl", void 0);
    Modal = __decorate([
        core_1.Component({
            selector: 'ig-modal',
            moduleId: module.id,
            template: "\n      <div #modal\n          tabindex=\"-1\"\n          [ngStyle]=\"{ display: isOpen ? 'block' : 'none' }\"\n          (click)=\"closeOnClick ? close() : 0\">\n          <div class=\"ig-modal__window\">\n              <ng-content></ng-content>\n          <div>\n      </div>\n    "
        }), 
        __metadata('design:paramtypes', [])
    ], Modal);
    return Modal;
}());
exports.Modal = Modal;
var ModalModule = (function () {
    function ModalModule() {
    }
    ModalModule = __decorate([
        core_1.NgModule({
            declarations: [Modal],
            exports: [Modal],
            imports: [common_1.CommonModule]
        }), 
        __metadata('design:paramtypes', [])
    ], ModalModule);
    return ModalModule;
}());
exports.ModalModule = ModalModule;
