// zero-blocks/input/input.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var IgInput = (function () {
    function IgInput(el) {
        this.el = el;
    }
    IgInput.prototype.onFocus = function (event) {
        this.focused = true;
    };
    IgInput.prototype.onBlur = function (event) {
        this.focused = false;
    };
    Object.defineProperty(IgInput.prototype, "filled", {
        get: function () {
            var value = this.el.nativeElement.value;
            return value && (value !== '');
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(IgInput.prototype, "placeholder", {
        get: function () {
            return this.el.nativeElement.getAttribute('placeholder') && !this.filled;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.HostListener('focus', ['$event']), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object]), 
        __metadata('design:returntype', void 0)
    ], IgInput.prototype, "onFocus", null);
    __decorate([
        core_1.HostListener('blur', ['$event']), 
        __metadata('design:type', Function), 
        __metadata('design:paramtypes', [Object]), 
        __metadata('design:returntype', void 0)
    ], IgInput.prototype, "onBlur", null);
    IgInput = __decorate([
        core_1.Directive({
            selector: '[igInput]',
            host: {
                '[class.ig-form-group__input]': 'true',
                '[class.ig-form-group__input--filled]': 'filled',
                '[class.ig-form-group__input--focused]': 'focused',
                '[class.ig-form-group__input--placeholder]': 'placeholder'
            }
        }), 
        __metadata('design:paramtypes', [core_1.ElementRef])
    ], IgInput);
    return IgInput;
}());
exports.IgInput = IgInput;
