// zero-blocks/checkbox/checkbox.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var noop = function () { };
var nextId = 0;
function MakeProvider(type) {
    return {
        provide: forms_1.NG_VALUE_ACCESSOR,
        useExisting: core_1.forwardRef(function () { return type; }),
        multi: true
    };
}
var IgCheckbox = (function () {
    function IgCheckbox() {
        this.id = "ig-checkbox-" + nextId++;
        this.disabled = false;
        this.tabindex = null;
        this.checked = false;
        this.change = new core_1.EventEmitter();
        this.focused = false;
        this._onTouchedCallback = noop;
        this._onChangeCallback = noop;
    }
    IgCheckbox.prototype.onChange = function (event) {
        if (this.disabled) {
            return;
        }
        this.checked = !this.checked;
        this._onChangeCallback(this.checked);
        this.change.emit(event);
    };
    IgCheckbox.prototype.onFocus = function (event) {
        this.focused = true;
    };
    IgCheckbox.prototype.onBlur = function (event) {
        this.focused = false;
        this._onTouchedCallback();
    };
    IgCheckbox.prototype.writeValue = function (value) {
        if (this.disabled) {
            return;
        }
        this._value = value;
        this.checked = !!this._value;
    };
    IgCheckbox.prototype.registerOnChange = function (fn) { this._onChangeCallback = fn; };
    IgCheckbox.prototype.registerOnTouched = function (fn) { this._onTouchedCallback = fn; };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IgCheckbox.prototype, "value", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], IgCheckbox.prototype, "id", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], IgCheckbox.prototype, "name", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], IgCheckbox.prototype, "disabled", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], IgCheckbox.prototype, "tabindex", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], IgCheckbox.prototype, "checked", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', Object)
    ], IgCheckbox.prototype, "change", void 0);
    __decorate([
        core_1.ViewChild('checkbox'), 
        __metadata('design:type', Object)
    ], IgCheckbox.prototype, "nativeCheckbox", void 0);
    IgCheckbox = __decorate([
        core_1.Component({
            selector: "ig-checkbox",
            moduleId: module.id,
            template: "\n      <div class=\"ig-control\">\n          <input class=\"ig-control__cbx\" #checkbox type=\"checkbox\"\n              [attr.id]=\"id\"\n              [attr.name]=\"name\"\n              [attr.value]=\"value\"\n              [attr.tabindex]=\"tabindex\"\n              [attr.disabled]=\"disabled ? true : null\"\n              [checked]=\"checked\"\n              (change)=\"onChange($event)\"\n              (focus)=\"onFocus($event)\"\n              (blur)=\"onBlur($event)\">\n          <label class=\"ig-control__label\" [attr.for]=\"id\">\n              <ng-content></ng-content>\n          </label>\n      </div>\n    ",
            providers: [MakeProvider(IgCheckbox)]
        }), 
        __metadata('design:paramtypes', [])
    ], IgCheckbox);
    return IgCheckbox;
}());
exports.IgCheckbox = IgCheckbox;
var CheckboxModule = (function () {
    function CheckboxModule() {
    }
    CheckboxModule = __decorate([
        core_1.NgModule({
            declarations: [IgCheckbox],
            exports: [IgCheckbox]
        }), 
        __metadata('design:paramtypes', [])
    ], CheckboxModule);
    return CheckboxModule;
}());
exports.CheckboxModule = CheckboxModule;
