// zero-blocks/core/touch.js
/// <reference path="../../../typings/globals/hammerjs/index.d.ts" />
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var EVENT_SUFFIX = "precise";
/**
 * Touch gestures manager based on Hammer.js
 * Use with caution, this will track references for single manager per element. Very TBD. Much TODO.
 */
var HammerGesturesManager = (function () {
    function HammerGesturesManager(_zone) {
        this._zone = _zone;
        this._hammerManagers = [];
        /**
         * Event option defaults for each recognizer, see http://hammerjs.github.io/api/ for API listing.
         */
        this.hammerOptions = [
            {
                name: "pan",
                options: {
                    threshold: 0
                }
            }, {
                name: "pinch",
                options: {
                    enable: true
                }
            }, {
                name: "rotate",
                options: {
                    enable: true
                }
            }];
    }
    HammerGesturesManager.prototype.supports = function (eventName) {
        return eventName.toLowerCase().endsWith("." + EVENT_SUFFIX);
    };
    /**
     * Add listener extended with options for Hammer.js. Will use defaults if none are provided.
     * Modeling after other event plugins for easy future modifications.
     */
    HammerGesturesManager.prototype.addEventListener = function (element, eventName, eventHandler, options) {
        if (options === void 0) { options = null; }
        var self = this;
        // Creating the manager bind events, must be done outside of angular
        return this._zone.runOutsideAngular(function () {
            // new Hammer is a shortcut for Manager with defaults
            var mc = new Hammer(element);
            for (var i = 0; i < self.hammerOptions.length; i++) {
                mc.get(self.hammerOptions[i].name).set(self.hammerOptions[i].options);
            }
            var handler = function (eventObj) { self._zone.run(function () { eventHandler(eventObj); }); };
            mc.on(eventName, handler);
            return function () { mc.off(eventName, handler); };
        });
    };
    /**
     * Add listener extended with options for Hammer.js. Will use defaults if none are provided.
     * Modeling after other event plugins for easy future modifications.
     *
     * @param target Can be one of either window, body or document(fallback default).
     */
    HammerGesturesManager.prototype.addGlobalEventListener = function (target, eventName, eventHandler) {
        var self = this, element = this.getGlobalEventTarget(target);
        // Creating the manager bind events, must be done outside of angular
        return this._zone.runOutsideAngular(function () {
            // new Hammer is a shortcut for Manager with defaults
            var mc = new Hammer(element);
            self.addManagerForElement(element, mc);
            for (var i = 0; i < self.hammerOptions.length; i++) {
                mc.get(self.hammerOptions[i].name).set(self.hammerOptions[i].options);
            }
            var handler = function (eventObj) {
                self._zone.run(function () {
                    eventHandler(eventObj);
                });
            };
            mc.on(eventName, handler);
            return function () { mc.off(eventName, handler); };
        });
    };
    /** temp replacement for DOM.getGlobalEventTarget(target) because DI won't play nice for now */
    HammerGesturesManager.prototype.getGlobalEventTarget = function (target) {
        switch (target) {
            case "window":
                return window;
            case "body":
                return document.body;
            default:
                return document;
        }
    };
    /**
     * Set HammerManager options.
     *
     * @param element The DOM element used to create the manager on.
     *
     * ### Example
     *
     * ```ts
     * manager.setManagerOption(myElem, "pan", { pointers: 1 });
     * ```
     */
    HammerGesturesManager.prototype.setManagerOption = function (element, event, options) {
        var manager = this.getManagerForElement(element);
        manager.get(event).set(options);
    };
    /**
     * Add an element and manager map to the internal collection.
     *
     * @param element The DOM element used to create the manager on.
     */
    HammerGesturesManager.prototype.addManagerForElement = function (element, manager) {
        this._hammerManagers.push({ element: element, manager: manager });
    };
    /**
     * Get HammerManager for the element or null
     *
     * @param element The DOM element used to create the manager on.
     */
    HammerGesturesManager.prototype.getManagerForElement = function (element) {
        var result = this._hammerManagers.filter(function (value, index, array) {
            return value.element === element;
        });
        return result.length ? result[0].manager : null;
    };
    /**
     * Destroys the HammerManager for the element, removing event listeners in the process.
     *
     * @param element The DOM element used to create the manager on.
     */
    HammerGesturesManager.prototype.removeManagerForElement = function (element) {
        var index = null;
        for (var i = 0; i < this._hammerManagers.length; i++) {
            if (element === this._hammerManagers[i].element) {
                index = i;
                break;
            }
        }
        if (index !== null) {
            var item = this._hammerManagers.splice(index, 1)[0];
            // destroy also
            item.manager.destroy();
        }
    };
    /** Destroys all internally tracked HammerManagers, removing event listeners in the process. */
    HammerGesturesManager.prototype.destroy = function () {
        for (var i = 0; i < this._hammerManagers.length; i++) {
            this._hammerManagers[i].manager.destroy();
        }
        this._hammerManagers = [];
    };
    HammerGesturesManager = __decorate([
        core_1.Injectable(), 
        __metadata('design:paramtypes', [core_1.NgZone])
    ], HammerGesturesManager);
    return HammerGesturesManager;
}());
exports.HammerGesturesManager = HammerGesturesManager;
