// zero-blocks/navbar/navbar.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var touch_1 = require("../core/touch");
var common_1 = require('@angular/common');
var button_1 = require("../button/button");
var Navbar = (function () {
    function Navbar(_location) {
        this._location = _location;
    }
    Navbar.prototype.canGoBack = function () {
        var isStackHisoryEmpty = window.history.length == 0;
        return isStackHisoryEmpty;
    };
    Navbar.prototype.navigateBack = function ($event) {
        this._location.back();
    };
    Navbar.prototype.hasIcon = function () {
        return this.icon != undefined;
    };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], Navbar.prototype, "title", void 0);
    __decorate([
        core_1.Input('icon'), 
        __metadata('design:type', String)
    ], Navbar.prototype, "icon", void 0);
    Navbar = __decorate([
        core_1.Component({
            selector: "ig-navbar",
            moduleId: module.id,
            template: "\n      <nav class=\"ig-navbar\">\n      \t<span class=\"ig-navbar__left-btn\" [class.ig-navbar__left-btn--disabled]=\"canGoBack() || !hasIcon()\"> \n              <button igButton=\"navbar\" (click)=\"navigateBack($event)\">\n                  <i class=\"material-icons\">{{ icon }}</i>\n              </button>\n          </span>\n      \t<span>\n              <h1 class=\"ig-navbar__title\">{{ title }}</h1>\n          </span>\n      \t<ng-content></ng-content>\n      </nav>\n    ",
            providers: [touch_1.HammerGesturesManager]
        }), 
        __metadata('design:paramtypes', [common_1.Location])
    ], Navbar);
    return Navbar;
}());
exports.Navbar = Navbar;
var NavbarModule = (function () {
    function NavbarModule() {
    }
    NavbarModule = __decorate([
        core_1.NgModule({
            imports: [button_1.ButtonModule],
            declarations: [Navbar],
            exports: [Navbar]
        }), 
        __metadata('design:paramtypes', [])
    ], NavbarModule);
    return NavbarModule;
}());
exports.NavbarModule = NavbarModule;
