// zero-blocks/layout/layout.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require('@angular/core');
var LayoutDirective = (function () {
    function LayoutDirective() {
        this.dir = "row";
        this.reverse = false;
        this.wrap = "nowrap";
        this.justify = "flex-start";
        this.itemAlign = "flex-start";
        this.display = 'flex';
    }
    Object.defineProperty(LayoutDirective.prototype, "flexwrap", {
        get: function () { return this.wrap; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayoutDirective.prototype, "justifycontent", {
        get: function () { return this.justify; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayoutDirective.prototype, "align", {
        get: function () { return this.itemAlign; },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(LayoutDirective.prototype, "direction", {
        get: function () {
            if (this.reverse) {
                return (this.dir == 'row') ? 'row-reverse' : 'column-reverse';
            }
            return (this.dir == 'row') ? 'row' : 'column';
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], LayoutDirective.prototype, "dir", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], LayoutDirective.prototype, "reverse", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], LayoutDirective.prototype, "wrap", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], LayoutDirective.prototype, "justify", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], LayoutDirective.prototype, "itemAlign", void 0);
    __decorate([
        core_1.HostBinding('style.display'), 
        __metadata('design:type', Object)
    ], LayoutDirective.prototype, "display", void 0);
    __decorate([
        core_1.HostBinding('style.flex-wrap'), 
        __metadata('design:type', Object)
    ], LayoutDirective.prototype, "flexwrap", null);
    __decorate([
        core_1.HostBinding('style.justify-content'), 
        __metadata('design:type', Object)
    ], LayoutDirective.prototype, "justifycontent", null);
    __decorate([
        core_1.HostBinding('style.align-content'), 
        __metadata('design:type', Object)
    ], LayoutDirective.prototype, "align", null);
    __decorate([
        core_1.HostBinding('style.flex-direction'), 
        __metadata('design:type', Object)
    ], LayoutDirective.prototype, "direction", null);
    LayoutDirective = __decorate([
        core_1.Directive({
            selector: '[layout]'
        }), 
        __metadata('design:paramtypes', [])
    ], LayoutDirective);
    return LayoutDirective;
}());
exports.LayoutDirective = LayoutDirective;
var FlexDirective = (function () {
    function FlexDirective() {
        this.grow = 1;
        this.shrink = 1;
        this.order = 0;
    }
    Object.defineProperty(FlexDirective.prototype, "style", {
        get: function () {
            return this.grow + " " + this.shrink;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(FlexDirective.prototype, "itemorder", {
        get: function () {
            return this.order || 0;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], FlexDirective.prototype, "grow", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], FlexDirective.prototype, "shrink", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], FlexDirective.prototype, "flex", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], FlexDirective.prototype, "order", void 0);
    __decorate([
        core_1.HostBinding('style.flex'), 
        __metadata('design:type', Object)
    ], FlexDirective.prototype, "style", null);
    __decorate([
        core_1.HostBinding('style.order'), 
        __metadata('design:type', Object)
    ], FlexDirective.prototype, "itemorder", null);
    FlexDirective = __decorate([
        core_1.Directive({
            selector: '[flex]'
        }), 
        __metadata('design:paramtypes', [])
    ], FlexDirective);
    return FlexDirective;
}());
exports.FlexDirective = FlexDirective;
var IgLayout = (function () {
    function IgLayout() {
    }
    IgLayout = __decorate([
        core_1.NgModule({
            declarations: [LayoutDirective, FlexDirective],
            exports: [FlexDirective, LayoutDirective]
        }), 
        __metadata('design:paramtypes', [])
    ], IgLayout);
    return IgLayout;
}());
exports.IgLayout = IgLayout;
