// zero-blocks/radio/radio.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var core_1 = require("@angular/core");
var forms_1 = require("@angular/forms");
var nextId = 0;
var noop = function () { };
function MakeProvider(type) {
    return {
        provide: forms_1.NG_VALUE_ACCESSOR,
        useExisting: core_1.forwardRef(function () { return type; }),
        multi: true
    };
}
var IgRadio = (function () {
    function IgRadio() {
        this.id = "ig-radio-" + nextId++;
        this.tabindex = null;
        this.change = new core_1.EventEmitter();
        this._onTouchedCallback = noop;
        this._onChangeCallback = noop;
    }
    IgRadio.prototype.onChange = function (event) {
        this.select();
        this.change.emit(event);
    };
    IgRadio.prototype.select = function () {
        this.checked = true;
        this._onChangeCallback(this.value);
    };
    IgRadio.prototype.writeValue = function (value) {
        this._value = value;
        this.checked = (this._value == this.value);
    };
    IgRadio.prototype.onFocus = function (event) {
        this.focused = true;
    };
    IgRadio.prototype.onBlur = function (event) {
        this.focused = false;
        this._onTouchedCallback();
    };
    IgRadio.prototype.registerOnChange = function (fn) { this._onChangeCallback = fn; };
    IgRadio.prototype.registerOnTouched = function (fn) { this._onTouchedCallback = fn; };
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Object)
    ], IgRadio.prototype, "value", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], IgRadio.prototype, "id", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', String)
    ], IgRadio.prototype, "name", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Boolean)
    ], IgRadio.prototype, "disabled", void 0);
    __decorate([
        core_1.Input(), 
        __metadata('design:type', Number)
    ], IgRadio.prototype, "tabindex", void 0);
    __decorate([
        core_1.Output(), 
        __metadata('design:type', core_1.EventEmitter)
    ], IgRadio.prototype, "change", void 0);
    __decorate([
        core_1.ViewChild('radio'), 
        __metadata('design:type', Object)
    ], IgRadio.prototype, "nativeRadio", void 0);
    IgRadio = __decorate([
        core_1.Component({
            selector: 'ig-radio',
            moduleId: module.id,
            template: "\n      <div class=\"ig-control\">\n          <input class=\"ig-control__radio\" type=\"radio\" #radio\n              [attr.id]=\"id\"\n              [attr.name]=\"name\"\n              [attr.value]=\"value\"\n              [attr.tabindex]=\"tabindex\"\n              [disabled]=\"disabled\"\n              [checked]=\"checked\"\n              (change)=\"onChange($event)\"\n              (focus)=\"onFocus($event)\"\n              (blur)=\"onBlur($event)\"/>\n          <label class =\"ig-control__label\" [attr.for]=\"id\"><ng-content></ng-content></label>\n      </div>\n    ",
            providers: [MakeProvider(IgRadio)]
        }), 
        __metadata('design:paramtypes', [])
    ], IgRadio);
    return IgRadio;
}());
exports.IgRadio = IgRadio;
var IgRadioModule = (function () {
    function IgRadioModule() {
    }
    IgRadioModule = __decorate([
        core_1.NgModule({
            declarations: [IgRadio],
            exports: [IgRadio]
        }), 
        __metadata('design:paramtypes', [])
    ], IgRadioModule);
    return IgRadioModule;
}());
exports.IgRadioModule = IgRadioModule;
